src/storage/mvcc/reader/reader.rs

```diff
@@ -939,4 +942,57 @@ mod tests {
         assert_eq!(write.write_type, WriteType::Put);
         assert_eq!(write.start_ts, 12);
     }
+
+    #[test]
+    fn test_check_lock() {
+        let path = Builder::new()
+            .prefix("_test_storage_mvcc_reader_check_lock")
+            .tempdir()
+            .unwrap();
+        let path = path.path().to_str().unwrap();
+        let region = make_region(1, vec![], vec![]);
+        let db = open_db(path, true);
+        let mut engine = RegionEngine::new(Arc::clone(&db), region.clone());
+
+        let (k1, k2, k3, v) = (b"k1", b"k2", b"k3", b"v");
+        engine.prewrite(Mutation::Put((Key::from_raw(k1), v.to_vec())), k1, 5);
+        engine.prewrite(Mutation::Put((Key::from_raw(k2), v.to_vec())), k1, 5);
+        engine.prewrite(Mutation::Lock(Key::from_raw(k3)), k1, 5);
+
+        let snap = RegionSnapshot::from_raw(Arc::clone(&db), region.clone());
+        let mut reader = MvccReader::new(snap, None, false, None, None, IsolationLevel::Si);
+        // Ignore the lock if read ts is less than the lock version
+        assert_eq!(reader.check_lock(&Key::from_raw(k1), 4).unwrap(), 4);
+        assert_eq!(reader.check_lock(&Key::from_raw(k2), 4).unwrap(), 4);
+        // Returns the lock if read ts >= lock version
+        assert!(reader.check_lock(&Key::from_raw(k1), 6).is_err());
+        assert!(reader.check_lock(&Key::from_raw(k2), 6).is_err());
+        // Read locks don't block any read operation
+        assert_eq!(reader.check_lock(&Key::from_raw(k3), 6).unwrap(), 6);
+        // Ignore the primary lock and returns the version before the lock
+        // when reading the latest committed version by setting u64::MAX as ts
+        assert_eq!(reader.check_lock(&Key::from_raw(k1), u64::MAX).unwrap(), 4);
+        // Should not ignore the secondary lock even though reading the latest version
+        assert!(reader.check_lock(&Key::from_raw(k2), u64::MAX).is_err());
+
+        // Commit the primary lock only
+        engine.commit(k1, 5, 7);
+        let snap = RegionSnapshot::from_raw(Arc::clone(&db), region.clone());
+        let mut reader = MvccReader::new(snap, None, false, None, None, IsolationLevel::Si);
+        // Then reading the primary key should succeed
+        assert_eq!(reader.check_lock(&Key::from_raw(k1), 6).unwrap(), 6);
+        // Reading secondary keys should still fail
+        assert!(reader.check_lock(&Key::from_raw(k2), 6).is_err());
+        assert!(reader.check_lock(&Key::from_raw(k2), u64::MAX).is_err());
+
+        // Pessimistic locks
+        let (k4, k5) = (b"k4", b"k5");
+        engine.acquire_pessimistic_lock(Key::from_raw(k4), k4, 9, 9);
+        engine.acquire_pessimistic_lock(Key::from_raw(k5), k5, 9, 9);
```

Comment:
There doesn't seem to be difference beteween k4 and k5. I think keep one of them is ok.